### ISSUE TITLE
kvclient/rangefeed: clients must handle internal errors

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/cockroach#75610]
Rangefeed clients based on `kvclient/rangefeed.Factory` must explicitly set an `OnInternalError` handler to be notified of internal errors. Otherwise, when one of these errors are encountered the rangefeed will silently stop emitting events. The current error conditions are:

* `roachpb.BatchTimestampBeforeGCError`: the rangefeed sent a batch request below the GC threshold.
* `roachpb.MVCCHistoryMutationError`: MVCC history was mutated by e.g. `AddSSTable`, `ClearRange`, or `RevertRange`.

Callers should add a callback using `WithOnInternalError` and handle these errors appropriately (as well as any future error conditions that may be added). Relevant callers include:

- [ ] [SQL descriptor watcher](https://github.com/cockroachdb/cockroach/blob/d5175f91c5d8d0bc547924d4b9c73761e93a09c8/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go#L249)
- [ ] [SQL zone watcher](https://github.com/cockroachdb/cockroach/blob/d5175f91c5d8d0bc547924d4b9c73761e93a09c8/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go#L306)
- [ ] [SQL lease watcher](https://github.com/cockroachdb/cockroach/blob/fa193c3277de67408c6507bf0e2ef036ab6f070b/pkg/sql/catalog/lease/lease.go#L1126)
- [ ] [SQL stats cache watcher](https://github.com/cockroachdb/cockroach/blob/98bd5a89b0d22ce0b764859a2b145bf2985e0bff/pkg/sql/stats/stats_cache.go#L170)
- [ ] [Settings watcher](https://github.com/cockroachdb/cockroach/blob/b08695e9160a479e265eee5e7631df76107484eb/pkg/server/settingswatcher/settings_watcher.go#L128)
- [ ] [Span config subscriber](https://github.com/cockroachdb/cockroach/blob/d5175f91c5d8d0bc547924d4b9c73761e93a09c8/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L282)

CC @cockroachdb/sql-schema @irfansharif 

Jira issue: CRDB-12744

**Comment by ajwerner** [comment]
This is legit. I have a PR to address this of for the settings and I think also the span configs. I'll revisit this later, as it's now easy enough to do with the pattern in #74612.

**Comment by ajwerner** [comment]
Turns out I missed the OnUnrecoverableError callback in #74612. Hat tip to @RaduBerinde for spotting it.

**Comment by ajwerner** [comment]
I'll note that at this point this is something, of a paranoia issue. Yes, we should do it, but it's not *that* pressing. There are two cases where these rangefeeds will fail in an unrecoverable way: the underlying span gets an AddSSTable or the GC TTL passes while the span is being watched. Those spans never get written to with AddSSTable, so that’s not really an issue. The GC TTL Is currently 25h, so these system tables would need to be unavailable for at least that long (though in practice, much longer) for this error to be hit.

Anyway, I do plan to do something here, just want to clarify the urgency.

**Comment by erikgrinaker** [comment]
I agree, this seems very unlikely -- I'm just a bit paranoid that some of these MVCC bulk ops changes can have unintended consequences, and silent failures are the worst kind of failure.

**Comment by rimadeodhar** [comment]
I'm going to move this over to the DB Server team for triage and execution as our team owns the sql liveness subsystem where this needs to be added. Additionally, I have tagged it with O-postmortem based on the discussion in https://cockroachlabs.slack.com/archives/C08NM67UWLX/p1745593841196779.
cc @cthumuluru-crdb

**Comment by rimadeodhar** [comment]
Actually, only the sql instance watcher work is relevant for DB Server. I have created a separate issue to track that https://github.com/cockroachdb/cockroach/issues/145356. 

Dropping the `O-postmortem` tag from this and moving this back to the SQL Foundations team.
